### PR TITLE
[SPARK-32668][SQL] HiveGenericUDTF initialize UDTF should use StructObjectInspector method

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

Use `initialize(StructObjectInspector argOIs)` instead `initialize(ObjectInspector[] args)` in `HiveGenericUDTF`.

### Why are the changes needed?

In our case, we implement a Hive `GenericUDTF` and override `initialize(StructObjectInspector argOIs)`. Then it's ok to execute with Hive, but failed with Spark SQL. Here is the Spark SQL error msg:
```
No handler for UDF/UDAF/UDTF 'com.xxxx.xxxUDTF': java.lang.IllegalStateException: 
Should not be called directly Please make sure your function overrides 
`public StructObjectInspector initialize(ObjectInspector[] args)`.
```

The reason is Spark `HiveGenericUDTF` call `initialize(ObjectInspector[] argOIs)` to init a UDTF, but it's a Deprecated method.
```
    public StructObjectInspector initialize(StructObjectInspector argOIs) throws UDFArgumentException {
        List<? extends StructField> inputFields = argOIs.getAllStructFieldRefs();
        ObjectInspector[] udtfInputOIs = new ObjectInspector[inputFields.size()];

        for(int i = 0; i < inputFields.size(); ++i) {
            udtfInputOIs[i] = ((StructField)inputFields.get(i)).getFieldObjectInspector();
        }

        return this.initialize(udtfInputOIs);
    }

    @Deprecated
    public StructObjectInspector initialize(ObjectInspector[] argOIs) throws UDFArgumentException {
        throw new IllegalStateException("Should not be called directly");
    }
```

We should use `initialize(StructObjectInspector argOIs)` to do this so that we can be compatible both of the two method. Same as Hive.

### Does this PR introduce _any_ user-facing change?

Yes, fix UDTF initialize method.

### How was this patch tested?

manual test and passed `HiveUDFDynamicLoadSuite`

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala [hunk 1]
```diff
@@ -210,10 +210,14 @@ private[hive] case class HiveGenericUDTF(
   }
 
   @transient
-  protected lazy val inputInspectors = children.map(toInspector)
+  protected lazy val inputInspectors = {
+    val inspectors = children.map(toInspector)
+    val fields = inspectors.indices.map(index => s"col$index").asJava
```
Comment (ulysses-you) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala:215]: The field name is not important so use col0 col1 ...

**Comment by ulysses-you** [comment]
The failed test caused by [#18527](https://github.com/apache/spark/pull/18527), I think we should support UTDF `UDTFStack2`. cc @wangyum

**Comment by dlutleixin** [comment]
yes, I have the same problem while using hive UDTF in spark-sql or spark.sql because of not override 
`public StructObjectInspector initialize(ObjectInspector[] args)`
then I override it according to error msg and it works.

**Comment by ulysses-you** [comment]
Then what do you think about this ? cc @dongjoon-hyun @sunchao

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @ulysses-you .
cc @gatorsmile , too.

**Comment by sunchao** [comment]
oops sorry @ulysses-you just remembered that you pinged me on this PR. This looks mostly good to me exception one question: since the new API is added in 0.13 and Spark [still support Hive 0.12](https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala#L92), do we need to take care of backward compatibility here?

cc @wangyum

**Comment by gatorsmile** [comment]
cc @somani

**Comment by ulysses-you** [comment]
thanks @sunchao . If don't miss something, Hive1.2 has been removed since [SPARK-32981](https://issues.apache.org/jira/browse/SPARK-32981) with branch-3.1. Hive0.12 is so far for master branch so we don't need care about compatible with it. cc @dongjoon-hyun isn't it ?

**Comment by gatorsmile** [comment]
@sunchao , you think this will be affected by the Hive metastore client versions?

**Comment by sunchao** [comment]
> @sunchao , you think this will be affected by the Hive metastore client versions?

This is the part I'm not sure, that is, Spark loading permanent UDF classes from a HMS with version 0.12. But apparently Hive 0.12 [doesn't support permanent UDF](https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala#L476) so seems this is not an issue.

**Comment by sunchao** [comment]
BTW #30665 is trying to solve the same issue and we should consolidate on one PR.

**Comment by wangyum** [comment]
ok to test.

**Comment by wangyum** [comment]
retest this please.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala [hunk 2]
```diff
@@ -210,10 +210,14 @@ private[hive] case class HiveGenericUDTF(
   }
 
   @transient
-  protected lazy val inputInspectors = children.map(toInspector)
+  protected lazy val inputInspectors = {
```
Comment (HyukjinKwon) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala:213]: nit: `inputInspectors` -> `inputInspector`

File: sql/hive/src/test/java/org/apache/spark/sql/hive/execution/UDTFStack3.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+/**
+ * Takes a row of size k of data and splits it into n rows of data.  For
+ * example, if n is 3 then the rest of the arguments are split in order into 3
+ * rows, each of which has k/3 columns in it (the first emitted row has the
+ * first k/3, the second has the second, etc).  If n does not divide k then the
+ * remaining columns are padded with NULLs.
+ */
+@Description(
+        name = "stack",
```
Comment (HyukjinKwon) [inline sql/hive/src/test/java/org/apache/spark/sql/hive/execution/UDTFStack3.java:39]: nit: 2 space indentation ..

**Comment by HyukjinKwon** [comment]
The code and PR description look making sense but I think somebody should take a close look ... I will leave it to @sunchao and @wangyum if you guys don't mind.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 4]
```diff
@@ -2176,8 +2176,7 @@ abstract class SQLQuerySuiteBase extends QueryTest with SQLTestUtils with TestHi
       sql(
         s"""
            |CREATE TEMPORARY FUNCTION udtf_stack2
-           |AS 'org.apache.spark.sql.hive.execution.UDTFStack2'
-           |USING JAR '${hiveContext.getHiveFile("SPARK-21101-1.0.jar").toURI}'
+           |AS 'org.apache.spark.sql.hive.execution.UDTFStack3'
```
Comment (wangyum) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala:2179]: Just replace this test with:
```scala
  test("SPARK-32668: HiveGenericUDTF initialize UDTF should use StructObjectInspector method") {
    withUserDefinedFunction("udtf_stack1" -> true, "udtf_stack2" -> true) {
      sql(
        s"""
           |CREATE TEMPORARY FUNCTION udtf_stack1
           |AS 'org.apache.spark.sql.hive.execution.UDTFStack'
           |USING JAR '${hiveContext.getHiveFile("SPARK-21101-1.0.jar").toURI}'
        """.stripMargin)
      sql(
        s"""
           |CREATE TEMPORARY FUNCTION udtf_stack2
           |AS 'org.apache.spark.sql.hive.execution.UDTFStack2'
           |USING JAR '${hiveContext.getHiveFile("SPARK-21101-1.0.jar").toURI}'
        """.stripMargin)

      Seq("udtf_stack1", "udtf_stack2").foreach { udf =>
        checkAnswer(
          sql(s"SELECT $udf(2, 'A', 10, date '2015-01-01', 'B', 20, date '2016-01-01')"),
          Seq(Row("A", 10, Date.valueOf("2015-01-01")),
            Row("B", 20, Date.valueOf("2016-01-01"))))
      }
    }
  }
```
?

`UDTFStack` and `UDTFStack2 ` are two ways to implement UDTF.

**Comment by sunchao** [comment]
I think there could be cases where one registers Hive 0.12 UDFs (although not sure how rare this is) in a HMS that Spark talks to, and things may fail when calling the UDFs since they don't have the new `initialize` method.

**Comment by ulysses-you** [comment]
@sunchao you mean user create a permanent udf which is from Hive0.12 build-in function ? If so I believe it's really rare ..

**Comment by sunchao** [comment]
@ulysses-you nvm please ignore my comment above :) I was thinking the case where Spark somehow loads the Hive 0.12 `GenericUDTF` class when executing user's custom UDTF but it seems that's not possible.

**Comment by wangyum** [comment]
Do we still need this error message?
https://github.com/apache/spark/blob/dfa6fb46f4238792bff6a0201da201be1b42620e/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala#L91-L93

**Review by wangyum — APPROVED** [review]

**Comment by wangyum** [comment]
Merged to master.

**Comment by ulysses-you** [comment]
thanks all!

**Pull request closed without merging**